proposals/public-suffix.md

```diff
@@ -0,0 +1,227 @@
+Mozilla's [multi-account-containers](https://github.com/mozilla/multi-account-containers)
+extension allows users to create containers for web pages as they visit them.
+
+Users may want the ability to apply the extension's functionality to a
+*wildcard domain*, such that all subdomains of some base domain are
+automatically included.
+
+This API allows the extension to automatically propose the topmost *registrable domain / eTLD+1*
+as a possible *wildcard domain* for the user to choose. It also allows the
+extension to prevent the user from accidentally choosing a *public suffix / eTLD*
+as a wildcard domain, e.g. `*.com`.
+
+#### Use Case #2: Grouping Domains in UI
+
+Where extensions present lists of domain names / URLs to users, it can be beneficial
+from a UX perspective to group them by their topmost *registrable domain / eTLD+1*s.
+
+##### Ungrouped domains
+
+|                       |
+| --------------------- |
+| example.co.uk         |
+| example2.com          |
+| foo.bar.example.co.uk |
+| foo.bar.example2.com  |
+| www.example.co.uk     |
+| www.example2.com      |
+
+##### Grouped domains
+
+| example.co.uk  |                       |
+| -------------- | --------------------- |
+|                | example.co.uk         |
+|                | foo.bar.example.co.uk |
+|                | www.example.co.uk     |
+
+| example2.com   |                       |
+| -------------- | --------------------- |
+|                | example2.com          |
+|                | foo.bar.example2.com  |
+|                | www.example2.com      |
+
+### Known Consumers
+
+Mozilla intends to make use of this API in its [multi-account-containers](https://github.com/mozilla/multi-account-containers) extension to allow users to create containers
+that include wildcard domain names: [PR #2352](https://github.com/mozilla/multi-account-containers/pull/2352). Currently, users must manually curate their containers to capture all
+anticipated subdomains of a base domain, and this manual approach is cumbersome and
+error-prone.
+
+Apart from this, any other extension that already rolls its own solution in order to parse
+the PSL could potentially benefit from this API, for example [uBlock Origin](https://github.com/gorhill/uBlock).
+
+## Specification
+
+### Schema
+
+A new API `publicSuffix` is added as follows:
+
+```ts
+//
+// Example:
+//
+// let domain = await browser.publicSuffix.getRegistrableDomain("www.example.co.uk");
+// ==> 'example.co.uk'
+//
+Promise<string> browser.publicSuffix.getRegistrableDomain(
+  // The domain name or URL whose registrable domain we want to find
+  domainNameOrUrl: string,
+)
+```
+
+### Behavior #1: Returned Promise
+
+The promise returned by `getRegistrableDomain()` method will either:
+
+1. Resolve with the topmost *registrable domain / eTLD+1* if it can be determined from
+the `domainNameOrUrl`.
+2. Reject with an `Error` / populate `browser.runtime.lastError`.
```

Comment:
Could you list examples that are explicitly considered valid or error conditions? The way that this 2-step behavior is phrased is very light on details and the lack of detail makes it more difficult for people to know what the API does and what it does not.

Some quick examples of specific classes of inputs:
- `example.net` (eTLD+1)
- `www.example.net` (eTLD+2)
- `net` (this is a public suffix itself)
- `127.0.0.1` (IP address, IPv4)
- `[::1]` (IPv6 address)
- `.example.net` (dot in front)
- `example.net.` (dot in the end, this is a FQDN)
- `github.io` (PSL entry marked as PRIVATE)
- `*.com` (listed elsewhere in this proposal as a use case, but an invalid domain name)

Some more examples are at https://url.spec.whatwg.org/#example-host-psl . These examples include IDN, non-canonical domains (e.g. upper case).

Reply:
Ok I will add this.

Reply:
Done 638833b2044492ca4b99415c1e5762e3c969d3e1